mqtt-streaming/src/main/scala/akka/stream/alpakka/mqtt/streaming/impl/ServerState.scala

```diff
@@ -0,0 +1,799 @@
+              val cc = ref.upcast[ClientConnection.Event]
+              cc ! ClientConnection.ConnectReceivedFromRemote(connect, local)
+              cc
+          }
+          context.watch(clientConnection)
+          val newConnection = (connectionId, (connect.clientId, clientConnection))
+          listening(
+            data.copy(
+              clientConnections = data.clientConnections
+                .filterNot { case (_, (clientId, _)) => clientId == connect.clientId } + newConnection
+            )
+          )
+        case (_, ConnAckReceivedLocally(connectionId, connAck, remote)) =>
+          forward(connectionId, data.clientConnections, ClientConnection.ConnAckReceivedLocally(connAck, remote))
+        case (_, SubscribeReceivedFromRemote(connectionId, subscribe, local)) =>
+          forward(connectionId, data.clientConnections, ClientConnection.SubscribeReceivedFromRemote(subscribe, local))
+        case (_, PublishReceivedFromRemote(connectionId, publish, local)) =>
+          forward(connectionId, data.clientConnections, ClientConnection.PublishReceivedFromRemote(publish, local))
+        case (_, PublishReceivedLocally(_, publish, publishData)) =>
+          data.clientConnections.values.foreach {
+            case (_, cc) => cc ! ClientConnection.PublishReceivedLocally(publish, publishData)
+          }
+          Behaviors.same
+        case (_, UnsubscribeReceivedFromRemote(connectionId, unsubscribe, local)) =>
+          forward(connectionId,
+                  data.clientConnections,
+                  ClientConnection.UnsubscribeReceivedFromRemote(unsubscribe, local))
+        case (_, PingReqReceivedFromRemote(connectionId, local)) =>
+          forward(connectionId, data.clientConnections, ClientConnection.PingReqReceivedFromRemote(local))
+        case (_, DisconnectReceivedFromRemote(connectionId, local)) =>
+          forward(connectionId, data.clientConnections, ClientConnection.DisconnectReceivedFromRemote(local))
+        case (_, ConnectionLost(connectionId)) =>
+          forward(connectionId, data.clientConnections, ClientConnection.ConnectionLost)
+      }
+      .receiveSignal {
+        case (_, t @ Terminated(ref)) =>
+          val terminatedCc = ref.upcast[ClientConnection.Event]
+          data.clientConnections.find { case (_, (_, cc)) => cc == terminatedCc } match {
+            case Some((connectionId, (clientId, _))) =>
+              if (t.failure.contains(ClientConnection.ClientConnectionFailed))
+                data.terminations.offer(ClientSessionTerminated(clientId))
+              listening(data.copy(clientConnections = data.clientConnections - connectionId))
+            case None =>
+              Behaviors.same
+          }
+      }
+}
+
+/*
+ * Handles events in relation to a specific client connection
+ */
+@InternalApi private[streaming] object ClientConnection {
+
+  /*
+   * No ACK received - the subscription failed
+   */
+  case object ClientConnectionFailed extends Exception
```

Comment:
I'd use `with NoStackTrace` here, since the stackTrace is questionable when sharing same exception instance with `object`

same in other places where object Exception is used

Reply:
Thanks!